instance/models/mixins/openedx_persistence.py

```diff
@@ -0,0 +1,182 @@
+class OpenEdXPersistenceMixin(MySQLInstanceMixin, MongoDBInstanceMixin, SwiftContainerInstanceMixin):
+    """
+    Mixin that provides functionality required for all the database and storage backends that an
+    OpenEdX Instance uses (when not using ephemeral databases)
+    """
+
+    @property
+    def database_name(self):
+        """
+        The database name used for external databases.
+        """
+        name = self.domain.replace('.', '_')
+        # Escape all non-ascii characters and truncate to 64 chars, the maximum for mysql:
+        allowed = string.ascii_letters + string.digits + '_'
+        escaped = ''.join(char for char in name if char in allowed)
+        return truncate_name(escaped, length=64)
+
+    @property
+    def mysql_database_name(self):
+        """
+        The mysql database name for this instance
+        """
+        return self.database_name
+
+    @property
+    def mysql_database_names(self):
+        """
+        List of mysql database names
+        """
+        return [self.mysql_database_name]
+
+    @property
+    def mongo_database_name(self):
+        """
+        The name of the main external mongo database
+        """
+        return self.database_name
+
+    @property
+    def forum_database_name(self):
+        """
+        The name of the external database used for forums
+        """
+        return '{0}_forum'.format(self.database_name)
+
+    @property
+    def mongo_database_names(self):
+        """
+        List of mongo database names
+        """
+        return [self.mongo_database_name, self.forum_database_name]
+
+    @property
+    def swift_container_name(self):
+        """
+        The name of the Swift container used by the instance.
+        """
+        return self.database_name
+
+    @property
+    def swift_container_names(self):
+        """
+        The list of Swift container names to be created.
+        """
+        return [self.swift_container_name]
+
+    def provision_mysql(self):
+        """
+        Set mysql credentials and provision the database.
+        """
+        if not self.mysql_provisioned:
+            self.mysql_user = get_random_string(length=16, allowed_chars=string.ascii_lowercase)
+            self.mysql_pass = get_random_string(length=32)
+        return super().provision_mysql()
+
+    def provision_mongo(self):
+        """
+        Set mongo credentials and provision the database.
+        """
+        if not self.mongo_provisioned:
+            self.mongo_user = get_random_string(length=16, allowed_chars=string.ascii_lowercase)
+            self.mongo_pass = get_random_string(length=32)
+        return super().provision_mongo()
+
+    def provision_swift(self):
+        """
+        Set Swift credentials and create the Swift container.
+        """
+        if settings.SWIFT_ENABLE and not self.swift_provisioned:
+            # TODO: Figure out a way to use separate credentials for each instance.  Access control
```

Comment:
@itsjeyd  I think in this case we only need separate credentials for each instance, not each AppServer. So I'll leave this as is.